pandas/core/arrays/sparse/array.py

```diff
@@ -945,13 +949,18 @@ def __getitem__(
             )
 
         else:
-            # TODO: I think we can avoid densifying when masking a
-            # boolean SparseArray with another. Need to look at the
-            # key's fill_value for True / False, and then do an intersection
-            # on the indices of the sp_values.
             if isinstance(key, SparseArray):
                 if is_bool_dtype(key):
-                    key = key.to_dense()
+                    if is_bool(key.fill_value):
```

Comment:
Here I'm not sure that such check is not redundant.  According to logic SparseDType(bool) should have bool fill_value, but in fact it could be an any value..

Reply:
there's an issue somewhere about being stricter about what fill_values are allowed for a given dtype, would be great to see an improvement there

Reply:
Hm..., actually right now I've added a changes to SparseDtype that will rise an exception in case of non boolean fill_value for SparseDtype(bool), but then I think what if user have an array with bool values and nan as fill_value? In this case we have to keep current functionality.